### DISCUSSION TITLE
How do I immediately update the value of an element in the middle of a function execution?

### DISCUSSION THREAD

**Discussion by ponponon** [zauberzeug/nicegui#2160] [Q&A]
### Question

How to update the value of an element immediately in the middle of function execution?

For example, in the following code, we want to show start first and then show done after 2 seconds.

But at the moment, it doesn't show start, it just shows done after two seconds of page blocking.

```python
import time
from nicegui import ui


def func(): with ui.dialog() as
    with ui.dialog() as dialog, ui.card().
        dialog.open()
        markdown = ui.markdown()
        markdown.content = 'start...'
        time.sleep(2)
        markdown.content = 'done'


ui.button('run', on_click=func)

ui.run(reload=True, port=8880)

```

How should I change the code to achieve the functionality I need?

**Comment by frankhuurman** [comment]
Your code seemed to have a few errors but I think I got the general idea of it so to answer your question, `ui.timer` is perfect for these kind of usecases since it doesn't block the script like `time.sleep`.

You could use this example to get the functionality you want:
```python
from nicegui import ui

def change_markdown(markdown):
    markdown.content = 'done'

def func():
    with ui.dialog() as dialog, ui.card():
        dialog.open()
        markdown = ui.markdown()
        markdown.content = 'start...'
        ui.timer(2.0, lambda: change_markdown(markdown), once=True)


ui.button('run', on_click=func)

ui.run(reload=True, port=8880)
```

**Comment by falkoschindler** [comment]
Alternatively to @frankhuurman's suggestion you can replace the synchronous `time.sleep()` with the asynchronous `asyncio.sleep()`:

```py
async def func():
    with ui.dialog() as dialog, ui.card():
        dialog.open()
        markdown = ui.markdown()
        markdown.content = 'start...'
        await asyncio.sleep(1)
        markdown.content = 'done'

ui.button('run', on_click=func)
```

In contrast to `time.sleep()`, which blocks the main thread until the time is up, `asyncio.sleep()` allows other tasks to be processed in the meantime. See [this FAQ](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-long-running-function-blocking-ui-updates) for more information.
